planetary-social/nos
# Remove SwiftGenPlugin and XCStringsToolPlugin

## Issues covered
#1117 

## Update June 25, 12:05 PM Eastern
The SwiftUI Previews build now works on its own! To verify:

1. Clean
2. Open a view with a preview, such as AboutView

It should build and show you a preview. It should also show around 832 warnings, all of which I guess were hidden when XCStringsToolPlugin was a plugin? I have a few things I'd still like to do:

- [ ] make SwiftGen output Assets.generated.swift instead of AssetsGenerated
- [x] make sure xcstringstool is all good with paths and script
- [ ] move script for SwiftGen from Xcode build phase into Scripts/
- [ ] move script for xcstrings-tool from Xcode build phase into Scripts/
- [x] should we move folders for tools into Scripts? probably
- [ ] fix warnings (how were these ignored before?)
- [x] ignore all Generated files in SwiftLint (why isn’t it already?) Localizable.generated shows line length, nesting, other violations
- [ ] consider using input files in Xcode so it doesn’t run xcstrings-tool every time (also re-check Based on dependency analysis)
- [ ] remove binary for SwiftGen
- [ ] remove binary for XCStringsTool (whoops!)

## Description
In the hope that this'll fix our build issue with SwiftGen previews, I've removed SwiftGenPlugin and replaced it with the SwiftGen binary. I also added the [SwiftGen-generated file](https://github.com/planetary-social/nos/pull/1261/files#diff-2b28627243162066459074ff0cd9c55482c576e4e18a3fb04b917affbd3a4066) to our repo since Matt and I discussed that (and since the other way just wouldn't compile for me).

Unfortunately, I still get errors like the following, which is the first of many:

```
== PREVIEW UPDATE ERROR:

    SchemeBuildError: Failed to build the scheme “Nos”
    
    type 'LocalizedStringResource' has no member 'moderation'
    
    Emitting module for NosTests:
    /Users/josh/Code/nos/Nos/Models/ReportCategory.swift:50:16: error: type 'LocalizedStringResource' has no member 'moderation'
            name: .moderation.coarseLanguage,
                  ~^~~~~~~~~~
```

I'm afraid this means that there's not an issue with SwiftGenPlugin; there's an issue with Xcode where using any plugin will cause the Preview build to fail. I'd love to be proven wrong, though, so please give it a shot.

## How to test
1. Clean and
4. Navigate to AboutView
5. Click the Refresh button on the Preview
6. Observe Preview

Ideally, the Preview would appear. However, at this point, the Preview build is still failing so you'll see a failure, and if you click the Diagnostics button you'll see the error above.

## Conclusion
I think that in order to get SwiftUI Previews to work when building them (and not require the dev to build the Test target), we have to remove both Swift Package Plugins. That would probably mean we need to either:

1. Use SwiftGen to generate the localization constants
2. Find a new tool to generate localization constants

In my brief search, neither seemed straightforward or simple. SwiftGen's documentation talks about generating code from `.strings` files, not `.xcstrings` files like we're using. Maybe someone out there has created a stencil for it, though.

I searched a bit for another tool, but all I found was the tool we're currently using: `XCStringsToolPlugin`.
